File: Override/calculater.rb
class Cal #객체 이름은 대문자여야 해
  attr_reader :v1, :v2
  attr_writer :v1
  @@_history = [] #클래스가 정의될 때 호출되는 배열. initialize가 불릴 때마다 기록할 것
  def initialize(v1, v2) #루비가 자동으로 실행하는 초기화 함수. instance가 실행될 때 꼭 실행돼. 뒤에 오는 애는 받아오는 매개변수
    p v1, v2 #initialize는 생성자.
    @v1 = v1 #@가 있는 변수는 instance 변수. 인스턴스에 소속된 모든 method에서 사용 가능.
    @v2 = v2 #@가 있으면, 그 instance 내 모든 method에서 접근 가능!
  end
#@가 없는 변수는 method 안에서만 쓸 수 있는 지역변수
#@ 안 달고 내가 initialze에만 name을 초기화한다든지 하면, 이 initialize에서만 쓸 수 있게 되는 거.
  def add()
    result = @v1 + @v2
    @@_history.push("add : #{v1} + #{v2} = #{result}") #push는 배열에 어떠한 값을 추가하는 method
    return result #샵 중괄호 해주면, 그 안에 있는 걸 변수로 인식하겠다는 뜻
  end
  def subtract()
    result = @v1 - @v2
    @@_history.push("subtract : #{v1} - #{v2} = #{result}") #push는 배열에 어떠한 값을 추가하는 method
    return result
  end
  def setV1(v)
    if v.is_a?(Integer)
      @v1 = v
    end
  end
  def getV1()
    return @v1
  end
  def Cal.history()
    for item in @@_history #히스토리 배열 안에 든 걸 하나씩 꺼내오겠다
      p item
    end
  end
  def info()
    return "Cal => v1 : #{@v1}, v2 : #{@v2}"
  end
end

class CalMultiply < Cal
  def multiply()
    result = @v1 * @v2
    @@_history.push("multiply : #{v1} x #{v2} = #{result}") #push는 배열에 어떠한 값을 추가하는 method
    return result
  end
  def info()
    return "CalMultiply => #{super()}"
  end
end

class CalDivide < CalMultiply
  def divide()
    result = @v1 / @v2
    @@_history.push("divide : #{v1} / #{v2} = #{result}") #push는 배열에 어떠한 값을 추가하는 method
    return result
  end
  def info()
    return "CalDivide => #{super()}"
  end 
end

c0 = Cal.new(30, 60)
p c0.info()
c1 = CalMultiply.new(10, 10)
p c1.info()
c2 = CalDivide.new(20, 10)
p c2.info()

File: Override/calculater.py
class Cal(object):
    _history = []
    def __init__(self, v1, v2): #언더바 두 개를 양 옆에, self를 안 넣고 그냥 v1 v2 하면 위치가 두 개 바뀌어서 들어감...!
#파이썬의 첫 번째 매개변수(self)는 꼭 정의해줘야 하고, 그 첫 번째 매개변수는 그 instance가 됨.
        if isinstance(v1, int): #검증된 값이 들어오게 하기 위해
            self.v1 = v1 #instance 변수로 만들기 위해
        if isinstance(v2, int):
            self.v2 = v2
    def add(self):
        result = self.v1 + self.v2
        Cal._history.append("add : %d+%d=%d" % (self.v1, self.v2, result)) #요소 추가하려면 append
        return result
    def subtract(self):
        result = self.v1 - self.v2
        Cal._history.append("subtract : %d-%d=%d" % (self.v1, self.v2, result)) #요소 추가하려면 append
        return result
    def setV1(self, v):
        if isinstance(v, int): #첫 번째 입력된 값이 두 번째 정의된 인자와 형식이 같다면 True 리턴
            self.v1 = v
    def getV1(self):
        return self.v1
    @classmethod
    def history(cls):
        for item in Cal._history:
            print(item)
    def info(self):
        return "Cal => v1 : %d, v2 : %d" % (self.v1, self.v2)

class CalMultiply(Cal):
    def multiply(self):
        result = self.v1 * self.v2
        Cal._history.append("multiply : %dx%d=%d" % (self.v1, self.v2, result)) #요소 추가하려면 append
        return result
    def info(self):
        return "CalMultiply => %s" % super().info()

#한 번 더 상속을 하고 싶을 때
class CalDivide(CalMultiply):
    def divide(self):
        result = self.v1 / self.v2
        Cal._history.append("divide : %d/%d=%d" % (self.v1, self.v2, result)) #요소 추가하려면 append
        return result
    def info(self):
        return "CalDivide => %s" % super().info()
#여기서 super는 CalMultiply

c0 = Cal(30, 60)
c1 = CalMultiply(10, 10)
c2 = CalDivide(20, 10)

print(c0.info())
print(c1.info())
print(c2.info())
